### PR TITLE
feat: task opt in and opt out

### PR DESCRIPTION
**Pull request by negar-abbasi**
## Proposed Changes

- opt in to multiple asset
- opt out on multiple assets, grouped into a single transaction 

## TODO
- [x] Add some docs
- [x] Add multiple assets opt in

### CODE REVIEW

File: src/asset.ts [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+import { TransactionGroupToSend, TransactionToSign } from './types/transaction'
+
+const MaxTxGroupSize = 16
+
+export async function optIn(client: Algodv2, account: Account, assetId: number) {
+  const accountInfo = await client.accountInformation(account.addr).do()
+  if (accountInfo.assets.find((a: Record<string, number>) => a['asset-id'] === assetId)) {
+    throw new Error(`Account ${account.addr} have already opt-in to asset ${assetId}`)
+  }
+  await transferAsset(
```
[the remark this answers is not in the document]
Comment (negar-abbasi) [reply src/asset.ts:13]: It is not hard but the user story is for only one asset to be opted in and a number of assets to be opted out.

File: src/asset.ts [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+import { Config, sendGroupOfTransactions } from '.'
+import { TransactionGroupToSend, TransactionToSign } from './types/transaction'
+
+const MaxTxGroupSize = 16
+
+function* chunks<T>(arr: T[], n: number): Generator<T[], void> {
+  for (let i = 0; i < arr.length; i += n) yield arr.slice(i, i + n)
+}
+
+async function ensureAccountIsValid(client: algosdk.Algodv2, account: algosdk.Account) {
```
Comment (aorumbayev) [inline src/asset.ts:11]: Same comment as in `py` version, are these meant to be used only by optin and optout or we want to export them? If we aren't exporting them ensure you are only exposing the optIn and optOut to users

File: src/asset.ts [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+  try {
+    await client.accountInformation(account.addr).do()
+  } catch (error) {
+    throw new Error(`Account address ${account.addr} does not exist`)
+  }
+}
+
+async function ensureAssetFirstOptIn(client: algosdk.Algodv2, account: algosdk.Account, assetIds: number[]) {
```
Comment (aorumbayev) [inline src/asset.ts:19]: Same comment as in `py` version, are these meant to be used only by optin and optout or we want to export them? If we aren't exporting them ensure you are only exposing the optIn and optOut to users

File: src/asset.ts [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        )
+      })
+    } catch (e) {
+      throw new Error(`Received error trying to opt in ${e}`)
+    }
+  }
+  return result
+}
+
+export async function optOut(client: Algodv2, optoutAccount: Account, assetIds: number[]): Promise<Record<number, string>> {
```
Comment (aorumbayev) [inline src/asset.ts:108]: in py version you named the account field as `account`, lets either rename it to account here or rename it to match the current title here on py version for consistency

Comment (aorumbayev) [inline src/asset.ts:108]:
```suggestion
/**
 * Opt out of multiple assets in Algorand blockchain.
 *
 * @param {Algodv2} client - An instance of the Algodv2 client used to interact with the Algorand blockchain.
 * @param {Account} optoutAccount - The Algorand account that wants to opt out of the assets.
 * @param {number[]} assetIds - An array of asset IDs representing the assets to opt out of.
 * @returns {Promise<Record<number, string>>} - A record object containing asset IDs as keys and their corresponding transaction IDs as values.
 */
export async function optOut(client: Algodv2, optoutAccount: Account, assetIds: number[]): Promise<Record<number, string>> {
```

File: src/asset.ts [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+  if (invalidAssets.length > 0) {
+    throw new Error(
+      `Assets ${invalidAssets.join(
+        ', ',
+      )} cannot be opted out. Ensure that they are valid and that the account has previously opted into them.`,
+    )
+  }
+}
+
+export async function optIn(client: Algodv2, optInAccount: Account, assetIds: number[]) {
```
Comment (aorumbayev) [inline src/asset.ts:63]:
```suggestion
/**
 * Opt in to a list of assets on the Algorand blockchain.
 *
 * @param client - An instance of the Algodv2 class from the `algosdk` library.
 * @param optInAccount - An instance of the Account class from the `algosdk` library representing the account that wants to opt in to the assets.
 * @param assetIds - An array of asset IDs that the account wants to opt in to.
 * @returns A record object where the keys are the asset IDs and the values are the corresponding transaction IDs for successful opt-ins.
 * @throws If there is an error during the opt-in process.
 */
export async function optIn(client: Algodv2, optInAccount: Account, assetIds: number[]) {
```

**Review by aorumbayev — CHANGES_REQUESTED** [review]
@negar-abbasi please also pull latest from main branch as well as its out of sync by a good # of commits

File: src/asset.ts [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+
+async function ensureAccountIsValid(client: algosdk.Algodv2, account: algosdk.Account) {
+  try {
+    await client.accountInformation(account.addr).do()
+  } catch (error) {
+    throw new Error(`Account address ${account.addr} does not exist`)
+  }
+}
+
+async function ensureAssetFirstOptIn(client: algosdk.Algodv2, account: algosdk.Account, assetIds: number[]) {
```
Comment (aorumbayev) [inline src/asset.ts:19]: This method and the one below are redundant, both are almost identical except for min balance parameters, you can combine them into a single method called ensureAssetBalance and take a third param of custom enum type call ValidationType (with option OptIn or OptOut), based on this enum you can them figure out whether to check if balance is >= 0 (opted in) or asset should not be inside accountInfo response at all (opted out)

**Review by aorumbayev — CHANGES_REQUESTED** [review]

**Review by aorumbayev — APPROVED** [review]

**Pull request merged**